b3dm/b3dm.go

```diff
@@ -0,0 +1,148 @@
+package b3dm
+
+import (
+	"bufio"
+	"encoding/binary"
+	"io"
+	"os"
+
+	"github.com/pkg/errors"
+	"github.com/qmuntal/gltf"
+)
+
+const (
+	B3DM_PROP_BATCH_LENGTH = "BATCH_LENGTH"
+	B3DM_PROP_RTC_CENTER   = "RTC_CENTER"
+)
+
+type B3dmHeader struct {
+	Magic                        [4]byte
+	Version                      uint32
+	ByteLength                   uint32
+	FeatureTableJSONByteLength   uint32
+	FeatureTableBinaryByteLength uint32
+	BatchTableJSONByteLength     uint32
+	BatchTableBinaryByteLength   uint32
+}
+
+func (h *B3dmHeader) GetSize() int64 {
+	return 28
+}
+
+func (h *B3dmHeader) GetByteLength() uint32 {
+	return h.ByteLength
+}
+
+func (h *B3dmHeader) GetFeatureTableJSONByteLength() uint32 {
+	return h.FeatureTableJSONByteLength
+}
+
+func (h *B3dmHeader) GetFeatureTableBinaryByteLength() uint32 {
+	return h.FeatureTableBinaryByteLength
+}
+
+func (h *B3dmHeader) GetBatchTableJSONByteLength() uint32 {
+	return h.BatchTableJSONByteLength
+}
+
+func (h *B3dmHeader) GetBatchTableBinaryByteLength() uint32 {
+	return h.BatchTableBinaryByteLength
+}
+
+type B3dmFeatureTable struct {
+	BatchLength int
+	RtcCenter   []float64
+}
+
+type B3dm struct {
+	Header       B3dmHeader
+	FeatureTable FeatureTable
+	BatchTable   BatchTable
+	Model        *gltf.Document
+}
+
+func B3dmFeatureTableDecode(header map[string]interface{}, buff []byte) map[string]interface{} {
+	ret := make(map[string]interface{})
+	l := getIntegerScalarFeatureValue(header, buff, B3DM_PROP_BATCH_LENGTH)
+	ret[B3DM_PROP_BATCH_LENGTH] = l
+	ret[B3DM_PROP_RTC_CENTER] = getFloat64Vec3FeatureValue(header, buff, B3DM_PROP_RTC_CENTER)
+	return ret
+}
+
+func B3dmFeatureTableEncode(header map[string]interface{}, data map[string]interface{}) []byte {
+	return nil
+}
+
+func (m *B3dm) GetFeatureTableView() B3dmFeatureTable {
+	ret := B3dmFeatureTable{}
+	ret.BatchLength = m.FeatureTable.Header[B3DM_PROP_BATCH_LENGTH].(int)
+	if m.FeatureTable.Header[B3DM_PROP_RTC_CENTER] != nil {
+		ret.RtcCenter = m.FeatureTable.Header[B3DM_PROP_RTC_CENTER].([]float64)
+	}
+	return ret
+}
+
+func (m *B3dm) GetHeader() Header {
+	return &m.Header
+}
+
+func (m *B3dm) GetFeatureTable() *FeatureTable {
+	return &m.FeatureTable
+}
+
+func (m *B3dm) GetBatchTable() *BatchTable {
+	return &m.BatchTable
+}
+
+type B3dmReader struct {
+	rs *bufio.Reader
+}
+
+func NewB3dmReader(r io.Reader) *B3dmReader {
+	return &B3dmReader{
+		rs: bufio.NewReader(r),
+	}
+}
+
+func (r *B3dmReader) DecodeHeader(d *B3dmHeader) error {
+	if err := binary.Read(r.rs, littleEndian, d); err != nil {
+		return errors.Wrap(err, "open failed for b3dm file")
```

Comment:
```suggestion
		return errors.Wrap(err, "failed to decode header")
```

This function does not treat files directly